File: codes/algorithm/find_ith_member.c
#include "sort/algorithm.h"
#include <stdio.h>
#include <stdlib.h>

static void shuffle(int *a, int n)
{
    int i,j;
    for(i=0; i<n; ++i) {
        j = random()%(n-2)+1;
        if (i % 2 == 0) {
            swap(&a[0], &a[j]);
        } else {
            swap(&a[n-1], &a[j]);
        }
    }
}

static int partition(int *a, int p, int r)
{
    int v = a[r];
    int i,j=p-1;

    for(i=p; i<r; ++i) {
        if (a[i] < v) {
            j++;
            swap(&a[i], &a[j]);
        }
    }
    swap(&a[j+1], &a[r]);
    return j+1;
}


int find_ith_number(int *a, int p, int r, int i)
{
    if (p == r) return a[p];

    int q = partition(a, p, r);
    int k = q-p+1;
    if (k == i) return a[q];
    else if (k > i)
        return find_ith_number(a, p, q-1, i);
    else
        return find_ith_number(a, q+1, r, i-k);
}

int main(int argc, char *argv[])
{
    int i=0;
    int n=20;
    int a[100];
    for(i=0; i<n; ++i) a[i] = i+1;
    shuffle(a, n);
    print_array(a,n);
    i=n/2;
    if (argc > 1) i = atoi(argv[1]);

    printf("find %d'th smallest number: %d\n",
           i, find_ith_number(a, 0, n-1, i));

    return 0;
}

File: learning/php/algorithm/find_ith_member.php
<?php

function swap(&$x, &$y) {
    $tmp = $x;
    $x = $y;
    $y = $tmp;
}

function partition(array &$inputs, $p, $r) {

    // randomized pivot
    $i = mt_rand($p, $r);
    swap($inputs[$i], $inputs[$r]);

    $v = $inputs[$r];
    $j = $p - 1;
    for($i=$p; $i<$r; $i++) {
        if ($inputs[$i] <= $v) {
            $j++;
            swap($inputs[$i], $inputs[$j]);
        }
    }

    swap($inputs[$r], $inputs[$j+1]);
    return $j+1;
}

/**
 * returns the i th smallest element of the array $inputs[p..r]
 */
function randomized_select(array &$inputs, $p, $r, $i) {
    if ($p == $r) {
        return $inputs[$p];
    }

    $q = partition($inputs, $p, $r);

    $k = $q - $p + 1;
    if ($i == $k) return $inputs[$q];
    else if ($i < $k) {
        return randomized_select($inputs, $p, $q-1, $i);
    } else {
        return randomized_select($inputs, $q+1, $r, $i-$k);
    }
}

$n = 20;
if ($argc > 1) $n = (int)$argv[1];
$inputs = range(1, $n);
shuffle($inputs);
printf("array(%d):\n%s\n", count($inputs), implode(" ", $inputs));

$i = mt_rand(1, $n);
printf("find %d'th smallest of array: ", $i);
print randomized_select($inputs, 0, $n-1, $i).PHP_EOL;

?>